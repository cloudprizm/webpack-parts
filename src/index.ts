export * from './parts'
export * from './utils'
export * from './rules'
export * from './shared'
export * from './loaders'
export * from './plugins'
export * from './factory'

export * from 'fp-ts/lib/Reader'