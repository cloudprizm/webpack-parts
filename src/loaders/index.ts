export * from './babel'
export * from './css'
export * from './sass'
export * from './typescript'
export * from './types'