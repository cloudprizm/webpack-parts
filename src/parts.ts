
import { array } from 'fp-ts/lib/Array'
import { Predicate } from 'fp-ts/lib/function'
import { reader, asks, Reader } from 'fp-ts/lib/Reader'
import { fromTraversable, Lens } from 'monocle-ts'

import { $PropertyType } from 'utility-types'
import { Configuration as WebpackConfig, Module, Plugin, Resolve, Rule, RuleSetRule } from 'webpack'
import { Configuration as DevServerConfiguration } from 'webpack-dev-server'

export type Entry = $PropertyType<WebpackConfig, 'entry'>
export type Externals = $PropertyType<WebpackConfig, 'externals'>
export type Targets = $PropertyType<WebpackConfig, 'target'>
export type Modes = NonNullable<$PropertyType<WebpackConfig, 'mode'>>
export type DevTool = $PropertyType<WebpackConfig, 'devtool'>
export type Output = $PropertyType<WebpackConfig, 'output'>
export type Optimization = $PropertyType<WebpackConfig, 'optimization'>
export type Rules = $PropertyType<Module, 'rules'>
export type Alias = $PropertyType<Resolve, 'alias'>
export type Extensions = $PropertyType<Resolve, 'extensions'>
export type MainFields = $PropertyType<Resolve, 'mainFields'>
export type RuleType = $PropertyType<Rule, 'type'>

export interface UserConfig {
  externals: Externals
  target: Targets,
  env: Modes
  dist: string
  port: number
  publicPath: string
  watch: boolean
}

export type Configuration = WebpackConfig & { devServer?: DevServerConfiguration }
export type ConfigDefinition<K = UserConfig> = Reader<Partial<K>, Configuration>
export type ChainableConfigDefinition<K = UserConfig> = (config: Configuration) => ConfigDefinition<K>
export type ConfigPart<A> = Reader<Partial<UserConfig>, A>
export type ChainableConfigPart = (config: Configuration) => ConfigDefinition

export const module = Lens.fromNullableProp<Configuration, Module, 'module'>('module', { rules: [] })
export const target = Lens.fromNullableProp<Configuration, Targets, 'target'>('target', 'web')
export const mode = Lens.fromNullableProp<Configuration, Modes, 'mode'>('mode', 'development')
export const devTool = Lens.fromNullableProp<Configuration, DevTool, 'devtool'>('devtool', 'cheap-eval-source-map')

export const externals = Lens.fromNullableProp<Configuration, Externals, 'externals'>('externals', {})
export const output = Lens.fromNullableProp<Configuration, Output, 'output'>('output', {})
export const resolve = Lens.fromNullableProp<Configuration, Resolve, 'resolve'>('resolve', {})
export const entry = Lens.fromNullableProp<Configuration, Entry, 'entry'>('entry', {})
export const optimization = Lens.fromNullableProp<Configuration, Optimization, 'optimization'>('optimization', {})

export const alias = Lens.fromNullableProp<Resolve, Alias, 'alias'>('alias', {})
export const extensions = Lens.fromNullableProp<Resolve, Extensions, 'extensions'>('extensions', [])
export const mainFields = Lens.fromNullableProp<Resolve, MainFields, 'mainFields'>('mainFields', [])

export const devServer = Lens.fromNullableProp<Configuration, DevServerConfiguration, 'devServer'>('devServer', {})
export const plugins = Lens.fromNullableProp<Configuration, Plugin[], 'plugins'>('plugins', [])
export const rules = Lens.fromNullableProp<Module, Rules, 'rules'>('rules', [])

// shorthands
// modifiers
export const withRules = module.compose(rules)
export const withExternals = externals
export const withOutput = output
export const withPlugins = plugins
export const withEntry = entry
export const withOptimization = optimization
export const withAlias = resolve.compose(alias)

export const makeConfig = <K = UserConfig>(config: Configuration): ConfigDefinition<K> => reader.of({ ...config })

// overridings
export const withTarget = target
export const withMode = mode
export const withDevTool = devTool
export const withExtensions = resolve.compose(extensions)
export const withMainFields = resolve.compose(mainFields)

// getters
export const getMode = mode

export const attachDevServer: ConfigDefinition = asks(env => ({
  contentBase: env.dist,
  historyApiFallback: true,
  publicPath: env.publicPath,
  compress: true,
  hotOnly: true,
  port: env.port,
  watchOptions: {
    poll: true
  }
}))

// predicates
type CheckUserConfig = Predicate<Partial<UserConfig>>

export const isLocal: CheckUserConfig = (userConfig) => !userConfig.env || userConfig.env === 'none'
export const isProduction: CheckUserConfig = (userConfig) => userConfig.env === 'production'
export const isDevelopment: CheckUserConfig = (userConfig) => userConfig.env === 'development'
export const isWeb: CheckUserConfig = (userConfig) => userConfig.target === 'web'
export const isNode: CheckUserConfig = (userConfig) => userConfig.target === 'node'

export { RuleSetRule, RuleSetLoader, Plugin, Module } from 'webpack'

// extra type
export type LoaderRule = ConfigPart<Rule>
export type PluginsRule = ConfigPart<Plugin[]>

// better would be to rename it to append
export const applyRuleAtTheEnd = (rule: Rule) => withRules.modify(existing => [...existing, rule])
export const applyRuleAtTheBeginning = (rule: Rule) => withRules.modify(existing => [rule, ...existing])

// make plural form
export const applyPluginAtTheEnd = (plugins: Plugin[]) => withPlugins.modify(existing => existing.concat(plugins))
export const applyPluginAtTheBeginning = (plugins: Plugin[]) => withPlugins.modify(existing => plugins.concat(existing))

export const ruleSetUse = Lens.fromProp<RuleSetRule, 'use'>('use')

export const rulesTraversal = fromTraversable(array)<RuleSetRule>()
export const traverseModuleRules = withRules.composeTraversal(rulesTraversal)