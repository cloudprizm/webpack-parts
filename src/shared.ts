import findUp from 'find-up'
import path from 'path'

const projectDir = process.cwd()
const rootDir = path.dirname(findUp.sync('.git') || process.cwd())
const node_modules = path.resolve(rootDir, 'node_modules')
const packageJSON = path.resolve(projectDir, 'package.json')
const projectSrc = path.resolve(projectDir, 'src')

export const paths = {
  project: projectDir,
  packageJSON,
  dist: path.resolve(projectDir, 'dist'),
  root: rootDir,
  node_modules,
  cache: path.join(node_modules, '.cache'),
}

export const getFileInProjectFolder = (file: string, start = projectDir) => path.resolve(projectDir, file)
export const getFileInSRCFolder = (file: string, start = projectSrc) => path.resolve(projectSrc, file)

export const defaultExtensions = ['.ts', '.tsx', '.js', '.mjs', '.json']

export const test = {
  code: /\.(t|j)sx?$/,
  node_modules: /node_modules/,
  sass: /\.s(a|c)ss$/,
}
