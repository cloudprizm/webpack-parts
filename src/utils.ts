import { addDevServerEntrypoints } from 'webpack-dev-server'
import { Configuration } from './parts'

export const addHMRScripts =
  (entry: string | string[], config: Configuration) => {
    const cfg = { entry: Array.isArray(entry) ? entry : [entry] }
    addDevServerEntrypoints(cfg, { ...config.devServer })
    return cfg.entry
  }