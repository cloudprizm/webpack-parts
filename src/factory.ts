import { asks, reader } from 'fp-ts/lib/Reader'

import {
  applyRuleAtTheEnd,
  ChainableConfigDefinition,
  ConfigPart,
  DevTool,
  isProduction,
  Modes,
  Output,
  withDevTool,
  withExtensions,
  withMainFields,
  withMode,
  withOutput,
} from './parts'

import {
  defaultSassRules,
  defaultCodeRules,
} from './rules'

import { defaultExtensions } from './shared'

const umdOutput = withOutput.set({
  globalObject: 'this',
  libraryTarget: 'umd',
  umdNamedDefine: true
})

const outputEnv: ConfigPart<Output> = asks(env => ({
  filename: isProduction(env) ? '[name].[contenthash].js' : '[name].js',
  chunkFilename: isProduction(env) ? '[name].[contenthash].chunk.js' : '[name].chunk.js',
  path: env.dist,
  publicPath: env.publicPath || ''
}))

export const withLibraryOutput = withOutput.modify(output => ({
  ...output,
  libraryTarget: 'commonjs2',
}))

const setExtensions = withExtensions.set(defaultExtensions)
const setMainFields = withMainFields.set(['browser', 'main', 'module'])

const setWebpackMode: ConfigPart<Modes> = asks(userConfig => userConfig.env || 'none')
const setDevTool: ConfigPart<DevTool> = asks(userConfig =>
  isProduction(userConfig)
    ? 'source-map'
    : 'cheap-eval-source-map'
)

export const basicConfig: ChainableConfigDefinition = config =>
  reader
    .of({ ...config })
    .map(umdOutput)
    .map(setExtensions)
    .map(setMainFields)
    .ap(setDevTool.map(withDevTool.set))
    .ap(setWebpackMode.map(withMode.set))
    .ap(outputEnv.map(partialOutput =>
      withOutput.modify(output => ({ ...output, ...partialOutput }))))

export const minimalTypescriptWithSass: ChainableConfigDefinition = config =>
  basicConfig(config)
    .ap(defaultSassRules.map(applyRuleAtTheEnd))
    .ap(defaultCodeRules.map(applyRuleAtTheEnd))
