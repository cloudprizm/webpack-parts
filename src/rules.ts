import {
  styleLoader,
  bulmaVariantsToDS,
  cssLoaderWithTypeGeneration,
  sassLoader,
  babelLoader,
  basicCSSLoader,
  tsLoader,
  LoadersBundle,
} from './loaders'

import { LoaderRule } from './parts'
import { ask, asks } from 'fp-ts/lib/Reader'
import { test } from './shared'

export const sassLibraryDevelopmentLoaders: LoadersBundle = input => [
  styleLoader(),
  bulmaVariantsToDS(),
  cssLoaderWithTypeGeneration(),
  sassLoader(),
]

export const sassDevelopmentLoaders: LoadersBundle = input => [
  styleLoader(),
  basicCSSLoader(),
  sassLoader(),
]

export type LazyLoaderRule = (bundle: LoadersBundle) => LoaderRule

export const makeSassRule: LazyLoaderRule =
  loadersBundle =>
    asks((userConfig) => ({
      test: test.sass,
      exclude: test.node_modules,
      use: loadersBundle(userConfig),
    }))

export const makeCodeRule: LazyLoaderRule =
  loadersBundle =>
    ask()
      .map((env) => ({
        test: test.code,
        exclude: test.node_modules,
        use: loadersBundle(env)
      }))

export const minimalCodeLoaders: LoadersBundle = input => [
  babelLoader(input.babelConfig),
  tsLoader()
]

export const defaultSassRules = makeSassRule(sassDevelopmentLoaders)
export const librarySassRules = makeSassRule(sassLibraryDevelopmentLoaders)
export const defaultCodeRules = makeCodeRule(minimalCodeLoaders)
