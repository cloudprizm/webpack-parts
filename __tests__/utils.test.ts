import { addHMRScripts } from '../src/utils'

test('adding hmr scripts to entry', () => {
  const devServer = {
    port: 9999,
    host: 'localhost',
    inline: true,
    hot: true
  }
  const decoratedEntryPoint = addHMRScripts('./src/some-entrypoint', { devServer })
  expect(decoratedEntryPoint).toMatchSnapshot()
})