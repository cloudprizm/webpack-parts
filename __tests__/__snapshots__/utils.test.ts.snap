// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding hmr scripts to entry 1`] = `
Array [
  "/Users/damianbaar/work/catch-the-sun/node_modules/webpack-dev-server/client/index.js?http://localhost:9999",
  "/Users/damianbaar/work/catch-the-sun/node_modules/webpack/hot/dev-server.js",
  "./src/some-entrypoint",
]
`;
